docs/tutorials/dataassociation/KDTree_Tutorial.py

```diff
@@ -0,0 +1,489 @@
+# detection, the search moves down the left subtree, and down the right subtree if the value is greater. This process is
+# repeated recursively at each internal node, again alternating through the vector dimensions, until a leaf node is
+# reached. See Figure 2 for example.
+#
+# To find the nearest neighbour to the prediction, at each node the Euclidean distance between the prediction and
+# the detection at that node is measured. If the distance is less than the current best distance, the detection at that
+# node becomes the new current best nearest neighbour.
+#
+# .. image:: ../../_static/kd_tree_fig_2.png
+#   :width: 1200
+#   :height: 250
+#   :alt: Figure showing search of k-d tree in 2 dimension
+#
+# Fig. 2a shows the detections (orange) and the prediction vector, :math:`[8,2]` (blue). 2b shows the search down the
+# :math:`k`-d tree with current best distance shown on the right side. We start at the root node detection, where
+# distance from prediction = 7.62 units. We look at the :math:`x^{th}` dimension of the prediction :math:`[8,2]` and
+# detection :math:`[5,9]` vectors. 8 > 5 so we move right down the tree to detection :math:`[8,5]` where distance from
+# prediction = 3. Now we look at the :math:`y^{th}` dimension of the vectors: 2 < 5 so we move left to the leaf node
+# containing detection :math:`[9,2]` where distance from prediction = 1. The search then continues back up the tree and
+# as there are no sibling subtrees that can beat the current best, the detection at :math:`[9,2]` is selected as the
+# nearest neighbour.
+
+# %%
+# Once we reach a leaf node the search must continue back up through the tree to ensure we haven't missed a potential
+# nearest neighbour that lies in an unexplored subtree. An unexplored subtree will only be searched if it could possibly
+# contain a closer detection than the current nearest neighbour.
+#
+# Looking at Figure 3 for example, after traversing completely down the tree, our current nearest neighbour to our
+# prediction vector :math:`[9,6]` is detection :math:`[7,5]` with a distance of 2.24 units. Figure 3a shows that the
+# shortest distance between the prediction (blue) and unexplored subtree (shaded area) is 1 unit (shown by red arrow).
+#
+# This means that there could be a detection in the unexplored subtree with a distance from prediction of between 1 and
+# 2.24 units, i.e. less than our current best of 2.24. So, we search down the left subtree from detection :math:`[7,5]`.
+# Indeed, there is a detection at :math:`[9,4]` in this subtree which has a distance from our prediction vector of 2
+# units, which becomes the new best distance. We then traverse back up the tree again. There are no other subtrees in
+# the tree which can possibly contain detections that beat the current best distance, so the detection at :math:`[9,4]`
+# is selected as the nearest neighbour and the search is complete when we reach the root.
+#
```

Comment:
One extra thought, what happens if both descendant nodes from the root node are an equal distance from the root node? Does the code account for this situation?